main.c

```diff
@@ -949,3 +959,126 @@ ErrorExit(int exit_code, const wchar_t *msg)
         exit(exit_code);
     }
 }
+
+/**
+ * Save a list of active connections in registry. If enable_auto_restart
+ * is false, any previously saved list is deleted and no new list is saved.
+ * Do not show any messages here as this may be called on receiving
+ * WM_ENDSESSION (user logout).
+ */
+static void
+SaveAutoRestartList()
+{
+    size_t len = 0;
+    int nactive = 0;
+    int max_active = o.num_configs - CountConnState(disconnected);
+
+    if (!o.enable_auto_restart || max_active <= 0)
+    {
+        /* delete the list -- on load this gets treated as an empty list */
+        RegDeleteKeyValueW(HKEY_CURRENT_USER, GUI_REGKEY_HKCU, L"auto_restart_list");
+        return;
+    }
+
+    int *active_conns = malloc((size_t) max_active*sizeof(int));
+    wchar_t *list = NULL;
+
+    if (!active_conns)
+    {
+        MsgToEventLog(EVENTLOG_ERROR_TYPE, L"Out of memory while persisting state in registry");
+        return;
+    }
+
+    for (int i = 0; i < o.num_configs; i++)
+    {
+        if (o.conn[i].state == disconnected
+            || o.conn[i].flags & FLAG_DAEMON_PERSISTENT)
+        {
+            continue;
+        }
+        /* accumulate space needed for list of active connections */
+        if (active_conns)
+        {
+            len += wcslen(o.conn[i].config_name) + 1;
+            active_conns[nactive++] = i;
+        }
+    }
+    len++; /* for double nul termination */
+
+    if (len == 1) len++; /* two nuls for empty string */
+
+    /* Make a double nul terminated list of active connections */
+    list = calloc(len, sizeof(wchar_t));
+    if (!list)
+    {
+        free(active_conns);
+        MsgToEventLog(EVENTLOG_ERROR_TYPE, L"Out of memory while persisting state in registry");
+        return;
+    }
+
+    wchar_t *p = list;
+    for (int i = 0; i < nactive; i++)
+    {
+        connection_t *c = &o.conn[active_conns[i]];
+        if (c->state == disconnected || c->flags & FLAG_DAEMON_PERSISTENT)
```

Comment:
We exclude disconnected and persistent connections in the loop above. Why do we need to check it here again?

Reply:
Right, no need to check persistent flag again. As for the state, it could have changed in the mean time in case the connection was in the process of disconnecting. But, on further thought, the second check is only marginally useful and its not critical to get this absolutely right. Will remove.